bench/various_appraches_bench.rb

```diff
@@ -0,0 +1,770 @@
+  )
+end
+
+def create_test_data(num_products:, num_variants:, num_selling_plan_groups:, num_selling_plans_per_group:, num_product_metafields: 0, num_variant_metafields: 0)
+  # Create SHARED selling plans - same instances used across all products
+  selling_plan_id = 1
+  selling_plan_groups = (1..num_selling_plan_groups).map do |group_idx|
+    selling_plans = (1..num_selling_plans_per_group).map do
+      plan = create_selling_plan(id: selling_plan_id)
+      selling_plan_id += 1
+      plan
+    end
+    create_selling_plan_group(id: group_idx, selling_plans: selling_plans)
+  end
+
+  # Create products that share the same selling plan groups
+  products = (1..num_products).map do |product_id|
+    create_product(
+      id: product_id,
+      num_variants: num_variants,
+      num_options: 2,
+      selling_plan_groups: selling_plan_groups,
+      num_product_metafields: num_product_metafields,
+      num_variant_metafields: num_variant_metafields,
+    )
+  end
+
+  products
+end
+
+# =============================================================================
+# Benchmark
+# =============================================================================
+
+SCENARIOS = [
+  # Baseline: minimal
+  { products: 1, variants: 1, product_metafields: 0, variant_metafields: 0, spg: 0, sp: 0 },
+
+  # Scale products
+  { products: 50, variants: 1, product_metafields: 0, variant_metafields: 0, spg: 0, sp: 0 },
+
+  # Scale variants
+  { products: 10, variants: 50, product_metafields: 0, variant_metafields: 0, spg: 0, sp: 0 },
+
+  # Scale product metafields
+  { products: 10, variants: 5, product_metafields: 20, variant_metafields: 0, spg: 0, sp: 0 },
+
+  # Scale variant metafields
+  { products: 10, variants: 5, product_metafields: 0, variant_metafields: 10, spg: 0, sp: 0 },
+
+  # Scale selling plans for ref_tracking
+  { products: 10, variants: 5, product_metafields: 0, variant_metafields: 0, spg: 1, sp: 2 },
+  { products: 10, variants: 5, product_metafields: 0, variant_metafields: 0, spg: 3, sp: 5 },
+
+  # Combined
+  { products: 10, variants: 10, product_metafields: 5, variant_metafields: 3, spg: 1, sp: 2 },
+  { products: 50, variants: 250, product_metafields: 20, variant_metafields: 10, spg: 3, sp: 5 },
+].freeze
+
+Report = Struct.new(:scenario, :bench_results, :bytesize_results, keyword_init: true)
+
+def run_benchmark(coders, scenario)
+  data = create_test_data(
+    num_products: scenario[:products],
+    num_variants: scenario[:variants],
+    num_selling_plan_groups: scenario[:spg],
+    num_selling_plans_per_group: scenario[:sp],
+    num_product_metafields: scenario[:product_metafields] || 0,
+    num_variant_metafields: scenario[:variant_metafields] || 0,
+  )
+
+  puts "\nBenchmarking scenario: P:#{scenario[:products]} V:#{scenario[:variants]} PM:#{scenario[:product_metafields]} VM:#{scenario[:variant_metafields]} SPG:#{scenario[:spg]} SP:#{scenario[:sp]}"
+
+  payloads = coders.map { |coder| coder.factory.dump(data) }
```

Comment:
```suggestion
  payloads = coders.map { |coder| coder.factory.dump(data).freeze }
```

FYI: If the payload is frozen, it saves a bunch of work for msgpack. The diff isn't massive, but on you simple benchmark it gives another ~0.10x  over marshal.